File: src/grounding/parse.py
"""
Shared document-text extraction.

Single source of truth for turning an uploaded PDF / DOCX / TXT into plain
text, reused by both the Streamlit UI and the FastAPI `/extract` endpoint.
"""
from __future__ import annotations

import io


def extract_text(filename: str, raw: bytes) -> str:
    """Extract plain text from a PDF, DOCX, or text file.

    Args:
        filename: original file name (used only to pick the parser by extension).
        raw: the raw file bytes.

    Returns:
        Extracted plain text. Unknown extensions are decoded as UTF-8.
    """
    name = filename.lower()
    if name.endswith(".pdf"):
        import pymupdf

        doc = pymupdf.open(stream=raw, filetype="pdf")
        return "\n\n".join(str(page.get_text()) for page in doc)
    if name.endswith(".docx"):
        from docx import Document as DocxDoc

        docx = DocxDoc(io.BytesIO(raw))
        return "\n".join(p.text for p in docx.paragraphs if p.text.strip())
    # Fall back to raw UTF-8 (txt, md, anything else)
    return raw.decode("utf-8", errors="replace")


def count_pdf_pages(raw: bytes) -> int | None:
    """Return the page count of a PDF, or None if it isn't a readable PDF."""
    try:
        import pymupdf

        with pymupdf.open(stream=raw, filetype="pdf") as doc:
            return doc.page_count
    except Exception:
        return None

File: dotnet/src/Grounding/Parse.cs
using System.Text;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using UglyToad.PdfPig;

namespace Grounding;

/// <summary>
/// Shared document-text extraction. Port of src/grounding/parse.py — turns an
/// uploaded PDF / DOCX / TXT into plain text. PyMuPDF → PdfPig, python-docx →
/// Open XML SDK.
/// </summary>
public static class Parse
{
    /// <summary>Extract plain text from a PDF, DOCX, or text file (by extension).</summary>
    public static string ExtractText(string filename, byte[] raw)
    {
        string name = filename.ToLowerInvariant();
        if (name.EndsWith(".pdf"))
        {
            using var doc = PdfDocument.Open(raw);
            return string.Join("\n\n", doc.GetPages().Select(p => p.Text));
        }
        if (name.EndsWith(".docx"))
        {
            using var stream = new MemoryStream(raw);
            using var word = WordprocessingDocument.Open(stream, false);
            var body = word.MainDocumentPart?.Document?.Body;
            if (body is null) return "";
            var lines = body.Descendants<Paragraph>()
                .Select(p => p.InnerText)
                .Where(t => t.Trim().Length > 0);
            return string.Join("\n", lines);
        }
        // Fall back to raw UTF-8 (txt, md, anything else).
        return Encoding.UTF8.GetString(raw);
    }

    /// <summary>Return the page count of a PDF, or null if it isn't a readable PDF.</summary>
    public static int? CountPdfPages(byte[] raw)
    {
        try
        {
            using var doc = PdfDocument.Open(raw);
            return doc.NumberOfPages;
        }
        catch
        {
            return null;
        }
    }
}
